src/Compilers/CSharp/Portable/Symbols/PublicModel/Symbol.cs

```diff
@@ -0,0 +1,245 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.PublicModel
+{
+    internal abstract class Symbol : ISymbol
+    {
+        internal abstract CSharp.Symbol UnderlyingSymbol { get; }
+
+
+        protected static ImmutableArray<TypeWithAnnotations> ConstructTypeArguments(ITypeSymbol[] typeArguments)
+        {
+            var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(typeArguments.Length);
+            foreach (var typeArg in typeArguments)
+            {
+                var type = typeArg.EnsureCSharpSymbolOrNull(nameof(typeArguments));
+                builder.Add(TypeWithAnnotations.Create(type, (typeArg?.NullableAnnotation.ToInternalAnnotation()).GetValueOrDefault()));
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        protected static ImmutableArray<TypeWithAnnotations> ConstructTypeArguments(ImmutableArray<ITypeSymbol> typeArguments, ImmutableArray<CodeAnalysis.NullableAnnotation> typeArgumentNullableAnnotations)
+        {
+            if (typeArguments.IsDefault)
+            {
+                throw new ArgumentException(nameof(typeArguments));
+            }
+
+            int n = typeArguments.Length;
+            if (!typeArgumentNullableAnnotations.IsDefault && typeArgumentNullableAnnotations.Length != n)
+            {
+                throw new ArgumentException(nameof(typeArgumentNullableAnnotations));
+            }
+
+            var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance(n);
+            for (int i = 0; i < n; i++)
+            {
+                var type = typeArguments[i].EnsureCSharpSymbolOrNull(nameof(typeArguments));
+                var annotation = typeArgumentNullableAnnotations.IsDefault ? NullableAnnotation.Oblivious : typeArgumentNullableAnnotations[i].ToInternalAnnotation();
+                builder.Add(TypeWithAnnotations.Create(type, annotation));
+            }
+
+            return builder.ToImmutableAndFree();
+        }
+
+        ISymbol ISymbol.OriginalDefinition
+        {
+            get
+            {
+                return UnderlyingSymbol.OriginalDefinition.GetPublicSymbol();
+            }
+        }
+
+        ISymbol ISymbol.ContainingSymbol
+        {
+            get
+            {
+                return UnderlyingSymbol.ContainingSymbol.GetPublicSymbol();
+            }
+        }
+
+        INamedTypeSymbol ISymbol.ContainingType
+        {
+            get
+            {
+                return UnderlyingSymbol.ContainingType.GetPublicSymbol();
+            }
+        }
+
+        public sealed override int GetHashCode()
+        {
+            return UnderlyingSymbol.GetHashCode();
```

Comment:
I'm not sure this is always correct. For the public `TypeSymbol`, I'd expect that the hashcode for this symbol would also include the `NullableAnnotation` of the symbol.

Reply:
> I'm not sure this is always correct. For the public TypeSymbol, I'd expect that the hashcode for this symbol would also include the NullableAnnotation of the symbol.

Including ```NullableAnnotation``` would be incorrect and we explicitly do not do this in all our hash code implementations for symbols. The rational - if symbols could be equal ignoring nullability, their hash code should be equal as well.

---
In reply to: [338815934](https://github.com/dotnet/roslyn/pull/39498#discussion_r338815934) [](ancestors = 338815934)